### PR TITLE
ports/nrf: Add WT51822-S4AT board.

### PR DESCRIPTION
**Pull request by aykevl**
I've added support for the [WT51822-S4AT](http://www.wireless-tag.com/wireless_module/BLE/WT51822-S4AT.html) board. I've tested most of it, but skipped I2C. SPI is currently disabled but could be enabled.

I bought the board on AliExpress so I don't know whether it's a real or a fake one (it's very cheap!), but it appears to work fine. It's known as nRF51822-04 there (also on eBay).

### CODE REVIEW

**Comment by glennrub** [comment]
Hi aykevl! Thanks a lot for the contribution.

The chip itself should is covered by the pca10001 dev-kit, which also uses a 2nd generation nrf51822 (nrf51822QFAAG0). However, the availability of pca10001 is quite limited now (cannot even find it on ebay), so i would say its worth getting support for WT51822-S4AT.

Could you elaborate a bit on what you have tested? 
Did you also run the Bluetooth SD on it? 
Did machine.ADC(5) on pin P0.04 work as expected?

I'll do a review of the PR and get back to you soon.

File: ports/nrf/boards/wt51822-s4at/mpconfigboard.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2016 Glenn Ruben Bakke
```
Comment (glennrub) [inline ports/nrf/boards/wt51822-s4at/mpconfigboard.h]: If you want you can use your name here. Also changing the year to 2017 as file is new.

File: ports/nrf/boards/wt51822-s4at/mpconfigboard.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+#define WT51822_S4AT
+
+// Datasheet for board:
+// https://4tronix.co.uk/picobot2/WT51822-S4AT.pdf
+#define MICROPY_HW_BOARD_NAME       "WT51822-S4AT"
+#define MICROPY_HW_MCU_NAME         "NRF51822"
+#define MICROPY_PY_SYS_PLATFORM     "nrf51"
+
+#define MICROPY_PY_MACHINE_HW_SPI   (0)
```
Comment (glennrub) [inline ports/nrf/boards/wt51822-s4at/mpconfigboard.h]: I suggest that we enable this one and adding the defines to the bottom of the pins used for SPI. It's hard to tell the usecase for everyone. If the SPI is not used, the pins are not occupied by any peripheral hardware.

config could look like this:
https://github.com/tralamazza/micropython/blob/master/ports/nrf/boards/microbit/mpconfigboard.h#L61-L65

When it comes to which pins to use its not many to choose from, but i suggest that staying away from P0.04 is a good guideline, which is the only routed analog pin (AIN5).

File: ports/nrf/boards/wt51822-s4at/nrf51_hal_conf.h [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+#define HAL_UART_MODULE_ENABLED
+#define HAL_TIME_MODULE_ENABLED
+#define HAL_RTC_MODULE_ENABLED
+#define HAL_TIMER_MODULE_ENABLED
+#define HAL_TWI_MODULE_ENABLED
+#define HAL_ADC_MODULE_ENABLED
+#define HAL_TEMP_MODULE_ENABLED
+#define HAL_RNG_MODULE_ENABLED
+
```
Comment (glennrub) [inline ports/nrf/boards/wt51822-s4at/nrf51_hal_conf.h]: If SPI is enabled, the HAL_SPI also has to be added into the compilation.

**Review by glennrub — CHANGES_REQUESTED** [review]
The PR looks good, with a few minor comments.
Some general feedback: 
- The folder should be renamed to wt51822_s4at with an underscore to keep it consistent with other boards. 
- The README.md should be updated with the the new board. (https://github.com/tralamazza/micropython/blob/master/ports/nrf/README.md#target-boards-and-make-flags). Maybe also the "tested hardware" section of the readme.

**Comment by aykevl** [comment]
Thank you for the feedback! I'll fix those things soon.
Some quick notes:
* I tested with (and without) the SoftDevice (s110). This was necessary for testing the Bluetooth console.
* I did some limited testing for the ADC, which happened to be P0.04. It appears to work but I'll do some proper testing (using a potmeter).
* It's actually a bit different from PCA10001: it doesn't have a 32kHz crystal and has a lot less pins.

I'll probably put SPI on pin 9, 10, 13: those pins are together on one side of the board.

**Comment by aykevl** [comment]
I've applied the changes as requested (via a force push)

`machine`:
* I tried to test I2C, but didn't get it to work - but that may also be caused by my hardware. I only tested `scan()` which always returns immediately with an empty list.
* I have enabled SPI, but haven't actually tested it (I don't think I have easy to test SPI hardware).
* ADC works, but appears to only show half of the range (between ~½Vcc and Vcc it always returns 255).
* For some reason, it appears like `machine.Pin` doesn't work over the BLE console. It works when using a serial connection, even while advertising via BLE (`ubluepy.Peripheral().advertise(device_name="abc")`). After I call `Pin.cpu.A4.init(mode=Pin.OUT, value=0)`, it doesn't respond anymore. I'm using the web console if that matters, as I couldn't get the NUS console to work.

File: ports/nrf/README.md [hunk 4]
```diff
@@ -80,6 +81,7 @@ pca10000             | s110                    | Peripheral             | [Segge
 pca10001             | s110                    | Peripheral             | [Segger](#segger-targets)
 pca10028             | s110                    | Peripheral             | [Segger](#segger-targets)
 pca10031             | s110                    | Peripheral             | [Segger](#segger-targets)
+wt51822-s4at         | s110                    | Peripheral             | [Segger](#segger-targets)
```
Comment (glennrub) [inline ports/nrf/README.md]: wt51822-s4at should be wt51822_s4at as the target name is very tied to the board folder name.

Also i wonder if we should set the "flash util" to something like "Manual debug port". As the device does not have an onboard debugger multiple debuggers could be used, but for all of them, it has to be manually connected.

**Comment by glennrub** [comment]
Great work! 

Yeah, as you said, it's a bit different than the pca10001. And also a very interesting device as it is quite limited on pinout :) 

I'm glad to hear that you have already tested the Bluetooth LE stack and that it is running quite well for you. I guess some of the problems you are seeing could also be related to the web console. You could try to send your commands with 20 characters at a time, to get the characteristic writes down to a MTU size, by pressing enter even in the middle of a line after 20 chars. 
The nus_console does not have these limitations, so it is totally worth trying to get it up running :-)

I'll do some testing tomorrow on the WebBluetooth console and see if i can reproduce the same issue here. 

Edit: 
When you say it does not respond any more, is the BLE connection also disconnected?

File: ports/nrf/README.md (hunk 4 above)
Comment (aykevl) [reply ports/nrf/README.md]: Fixed.

I've personally used (a clone of) the ST-Link v2 together with OpenOCD, which works well. Unfortunately, that means I had to flash the firmware manually (I didn't try the OpenOCD support included).

**Comment by aykevl** [comment]
> When you say it does not respond any more, is the BLE connection also disconnected?

I'm not sure. There came no reply, and when I sent more commands after that there was an error in Developer Tools complaining it's still busy handling (sending?) the previous command. Only a device reset fixes the problem (IIRC simply reloading the page and reconnecting wouldn't work - the device disappears).
If that's relevant, I'm using Linux (Debian stretch). My laptop is a Lenovo L450.

**Comment by glennrub** [comment]
I did some testing using web console and pca10001. It looks like it matters whether its running in Chrome in Ubuntu or Android. On my tablet, it is working correctly, but on ubuntu i cannot write to the RX characteristic of the NUS service. I have raised an issue for this (#108) . I suspect it is due to some event not handled correctly in the ble_drv.c when using s110 softdevice. When using s132 softdevice on nrf52832 it works in both ubuntu/android. 

However, i believe this should not block this PR as this is not related to the board support.

**Review by glennrub — APPROVED** [review]

**Pull request merged**